sourcerer-core/src/main/java/org/elder/sourcerer2/DefaultCommand.kt

```diff
@@ -0,0 +1,235 @@
+import org.elder.sourcerer2.exceptions.UnexpectedVersionException
+import org.slf4j.LoggerFactory
+import java.util.ArrayList
+import java.util.HashMap
+
+/**
+ * Default Command implementation, expressed in terms of an AggregateRepository.
+ */
+class DefaultCommand<TState, TParams, TEvent>(
+        private val repository: AggregateRepository<TState, TEvent>,
+        private val operation: Operation<TState, TParams, TEvent>
+) : Command<TState, TParams, TEvent> {
+    private val metadata: MutableMap<String, String>
+    private var atomic = true
+    private var idempotentCreate = false
+    private var aggregateId: StreamId? = null
+    private var arguments: TParams? = null
+    private var expectedVersion: ExpectedVersion? = null
+    private val metadataDecorators: MutableList<MetadataDecorator>
+
+    init {
+        Preconditions.checkNotNull(repository)
+        Preconditions.checkNotNull(operation)
+        this.atomic = operation.atomic()
+        this.metadata = HashMap()
+        this.metadataDecorators = ArrayList()
+    }
+
+    override fun setAggregateId(aggregateId: StreamId): Command<TState, TParams, TEvent> {
+        this.aggregateId = aggregateId
+        return this
+    }
+
+    override fun setArguments(arguments: TParams): Command<TState, TParams, TEvent> {
+        this.arguments = arguments
+        return this
+    }
+
+    override fun setExpectedVersion(version: ExpectedVersion): Command<TState, TParams, TEvent> {
+        this.expectedVersion = version
+        return this
+    }
+
+    override fun setAtomic(atomic: Boolean): Command<TState, TParams, TEvent> {
+        this.atomic = atomic
+        return this
+    }
+
+    override fun setIdempotentCreate(idempotentCreate: Boolean): Command<TState, TParams, TEvent> {
+        this.idempotentCreate = idempotentCreate
+        this.atomic = true
+        return this
+    }
+
+    override fun addMetadata(metadata: Map<String, String>): Command<TState, TParams, TEvent> {
+        this.metadata.putAll(metadata)
+        return this
+    }
+
+    override fun addMetadataDecorator(
+            metadataDecorator: MetadataDecorator): Command<TState, TParams, TEvent> {
+        this.metadataDecorators.add(metadataDecorator)
+        return this
+    }
+
+    override fun validate(): Command<TState, TParams, TEvent> {
+        if (aggregateId == null) {
+            throw InvalidCommandException("No aggregate id specified")
+        }
+        if (operation.requiresArguments() && arguments == null) {
+            throw InvalidCommandException(
+                    "No arguments specified to command that requires arguments")
+        }
+
+        // Try to calculate effective expected version - will validate combinations
+        getEffectiveExpectedVersion(expectedVersion, operation.expectedVersion())
+
+        // TODO: Add more validation!
+        return this
+    }
+
+    override fun run(): CommandResult<TEvent> {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Prefer splitting up complex methods into smaller, easier to understand methods. MCC 13 is greater than the threshold 10. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.elder.org/coding_rules#rule_key=detekt-kotlin%3AComplexMethod)